lib/ecto/repo/schema.ex

```diff
@@ -711,7 +727,15 @@ defmodule Ecto.Repo.Schema do
         {{:nothing, [], conflict_target}, []}
 
       {:replace, keys} when is_list(keys) ->
-        fields = Enum.map(keys, &field_source!(schema, &1))
+        fields =
+          Enum.map(keys, fn key ->
+            unless writable_field?(schema, key) do
+              raise ArgumentError, "cannot replace unwritable field `#{key}` in :on_conflict option"
+            end
```

Comment:
Weird question but, do we need this check? I assume we are not going to send the field anyway, so something else will crash?

Reply:
I was hoping for that too but the only place it crashes is here:

```elixir
field_source!(schema, key)
```

And we'd need to support field sources for read only fields.

Reply:
The only other thing I can think is if we make `writable_field_source` similar to how we made `writable_type`.

Reply:
I was going to say maybe I can put read_only? as part of the current field_source output but it's public so I believe that's out.